README
WWW::Shorten

   A Perl interface to URL shortening sites. These sites maintain
   databases of long URLs, each of which has a unique identifier.

For installation details, see INSTALL.

For usage, see the doco. Either use perldoc or view it
on the web at:
    <http://search.cpan.org/author/SPOON/WWW-Shorten/>

Changes for 1.5.7 (10 Mar 2003)
   - Modules return undef when unable to shorten.
   - Added link to NotLong's comparison page.
   - Revised some BUGS sections.
   - MASL returns URL if too short, else undef.
   - MASL tests extended.

For a complete list of changes, see the Changes file.


by Iain Truskett.


Changes
www-shorten (0.1.1) unstable; urgency=low

  * bin/shorten, lib/WWW/Shorten.pm, lib/WWW/Shorten/Metamark.pm,
    t/00load_a.t, t/00load_b.t, t/00noexp.t, t/metamark.t,
    t/newnames.t: Metamark is dead :-(

 -- Dave Cross <dave@dave.org.uk>  Mon, 15 Sep 2014 21:30:35 +0100

www-shorten (0.1.1) unstable; urgency=low

  * lib/WWW/Shorten.pm: Typo in documentation.
  * lib/WWW/Shorten.pm: Bump version for CPAN release.
  * INSTALL: Updated the INSTALL instructions.
  * t/tinyurl.t: (Possible) fix to the TinyURL test.
  * README: Modernised README
  * .travis.yml: Added .travis.yml

 -- Dave Cross <dave@dave.org.uk>  Tue, 6 May 2014 15:50:50 +0100

www-shorten (0.1.1) unstable; urgency=low

  * MANIFEST: Changes to reflect moved files.
  * t/98pod-coverage.t, t/99pod.t, xt/98pod-coverage.t, xt/99pod.t:
    Moved Pod tests into xt. (Pod coverage currently fails on some
    files. See https://rt.cpan.org/Ticket/Display.html?id=87634 for
    details.)

 -- Dave Cross <dave@dave.org.uk>  Thu, 22 Aug 2013 12:00:16 +0100

www-shorten (0.1.1) unstable; urgency=low

  * lib/WWW/Shorten.pm: Bump version number for release.
  * lib/WWW/Shorten/Linkz.pm, lib/WWW/Shorten/Metamark.pm,
    lib/WWW/Shorten/TinyURL.pm, lib/WWW/Shorten/Tinylink.pm,
    t/linkz.t, t/metamark.t, t/newnames.t, t/tinyurl.t: Removed
    prototypes (and the ampersands in the tests that circumvented
    them)

 -- Dave Cross <dave@dave.org.uk>  Mon, 12 Nov 2012 16:40:29 +0000

www-shorten (0.1.1) unstable; urgency=low

  * Artistic, Build.PL, COPYING, LICENCE, MANIFEST,
    lib/WWW/Shorten.pm, lib/WWW/Shorten/UserAgent.pm: Licensing
    clean-up.
  * MANIFEST: Added META.json to MANIFEST.

 -- Dave Cross <dave@dave.org.uk>  Fri, 6 Apr 2012 17:40:42 +0100

www-shorten (0.1.1) unstable; urgency=low

  * lib/WWW/Shorten/UserAgent.pm: Added documentation.
  * MANIFEST, lib/WWW/Shorten/UserAgent.pm,
    lib/WWW/Shorten/generic.pm: Added a WWW::Shorten::UserAgent object
    which dies if it receives an HTTP error response.

 -- Dave Cross <dave@dave.org.uk>  Thu, 1 Mar 2012 21:49:10 +0000

www-shorten (0.1.1) unstable; urgency=low

  * bin/shorten: Better examples of using bin/shorten
  * lib/WWW/Shorten.pm: Bump version number for release.
  * bin/shorten: Be far more intelligent about the code that allows
    the user to choose which service to use.
  * MANIFEST.SKIP: Added MYMETA.yml to MANIFEST.SKIP.
  * bin/shorten: Default to using a service that we currently support.
  * Build.PL, lib/WWW/Shorten.pm: Added Config::Auto to list of
    dependencies (it's used by the shorten program). Bumped version
    for release.

 -- Dave Cross <dave@dave.org.uk>  Mon, 6 Dec 2010 12:35:43 +0000

www-shorten (0.1.1) unstable; urgency=low

  * Build.PL: Removed version number so it's picked up from
    lib/WWW/Shorten.pm

 -- Dave Cross <dave@dacross.(none)>  Wed, 20 Jan 2010 16:21:06 +0000

www-shorten (0.1.1) unstable; urgency=low

  * Build.PL, MANIFEST: Bumped version number. Removed shorl files
    from MANIFEST.
  * MANIFEST, lib/WWW/Shorten/Shorl.pm, t/shorl.t: Removed support for
    shorl.pm (now in WWW::Shorten::Shorl distribution).

 -- Dave Cross <dave@dave.org.uk>  Sat, 19 Dec 2009 15:44:58 +0000

www-shorten (0.1.1) unstable; urgency=low

  * lib/WWW/Shorten/Shorl.pm: shorl.com was change the request method
    ( POST to GET )

 -- yappo <yappo@shibuya.pl>  Tue, 15 Dec 2009 11:46:14 +0800

www-shorten (0.1.1) unstable; urgency=low

  * lib/WWW/Shorten/NotLong.pm, lib/WWW/Shorten/OneShortLink.pm,
    t/notlong.t, t/oneshort.t: Removed support for NotLong and
    OneShortLink (separate distributions to follow soon). Bumper to
    version 2.06.

 -- Dave Cross <dave@angel.mag-sol.com>  Sat, 21 Nov 2009 21:36:14 +0000

www-shorten (0.1.1) unstable; urgency=low

  * Build.PL: Bump to version 2.05 for release.
  * lib/WWW/Shorten.pm, t/newnames.t: Finish removing all references
    to SnipURL.pm.

 -- Dave Cross <dave@dave.org.uk>  Fri, 18 Sep 2009 13:46:00 +0100

www-shorten (0.1.1) unstable; urgency=low

  * MANIFEST: Removed snipurl.t from MANIFEST.
  * Build.PL, MANIFEST, lib/WWW/Shorten/SnipURL.pm, t/snipurl.t:
    Removed WWW::Shorten::SnipURL (API changed, separate dist will
    follow). Bumped version to 2.04.

 -- Dave Cross <dave@dave.org.uk>  Tue, 8 Sep 2009 11:54:17 +0100

www-shorten (0.1.1) unstable; urgency=low

  * lib/WWW/Shorten.pm: Start using $DEFAULT_SERVICE variable.
  * MANIFEST: Removed files from MANIFEST
  * lib/WWW/Shorten.pm: Documenation improvements.
  * lib/WWW/Shorten/BabyURL.pm, lib/WWW/Shorten/EkDk.pm,
    lib/WWW/Shorten/ShortLink.pm, lib/WWW/Shorten/SmLnk.pm,
    lib/WWW/Shorten/URLjr.pm, lib/WWW/Shorten/V3.pm,
    lib/WWW/Shorten/qURL.pm, t/babyurl.t, t/ekdk.t, t/qurl.t,
    t/shortlink.t, t/smlnk.t, t/urljr.t, t/v3.t: Removed a number of
    modules that supported services that are no longer around.

 -- Dave Cross <dave@dave.org.uk>  Sat, 11 Apr 2009 12:51:19 +0100

www-shorten (0.1.1) unstable; urgency=low

  * Build.PL: Remove signing
  * t/shorl.t: Fixed typo
  * Build.PL: Bump version number.
  * t/tinyurl.t: Fixed TinyURL tests.

 -- Dave Cross <dave@dave.org.uk>  Thu, 19 Mar 2009 10:32:23 +0000

www-shorten (0.1.1) unstable; urgency=low

  * Build.PL: Increment version number

 -- Dave Cross <dave@dave.org.uk>  Mon, 2 Feb 2009 15:28:53 +0000

www-shorten (0.1.1) unstable; urgency=low

  * Build.PL: Changed require perl 5.6.0 to require perl 5.006 - to
    deal with older Perl installations.

 -- Dave Cross <dave@dave.org.uk>  Mon, 26 Jan 2009 13:27:29 +0000

www-shorten (0.1.1) unstable; urgency=low

  * Build.PL: Increment version
  * MANIFEST: Removed SIGNATURE from MANIFEST (as Module::Build adds
    it automatically).
  * lib/WWW/Shorten/Tinylink.pm: Typo

 -- Dave Cross <dave@dave.org.uk>  Mon, 30 Jun 2008 16:01:36 +0000

www-shorten (0.1.1) unstable; urgency=low

  * Build.PL, lib/WWW/Shorten/Tinylink.pm, t/tinylink.t: Marked

lib/WWW/Shorten.pm
#$Id: Shorten.pm 44 2007-04-15 14:20:35Z dave $

=head1 NAME

WWW::Shorten - Interface to URL shortening sites.

=head1 SYNOPSIS

  use WWW::Shorten 'BabyURL';
  use WWW::Shorten 'Linkz';
  use WWW::Shorten 'MakeAShorterLink';
  use WWW::Shorten 'Metamark';
  use WWW::Shorten 'NotLong';
  use WWW::Shorten 'OneShortLink';
  use WWW::Shorten 'Shorl';
  use WWW::Shorten 'SnipURL';
  use WWW::Shorten 'TinyClick';
  use WWW::Shorten 'TinyLink';
  use WWW::Shorten 'TinyURL';

  # These 6 are now inactive and will give an
  # error if you try to use them.
  use WWW::Shorten 'EkDk';
  use WWW::Shorten 'qURL';
  use WWW::Shorten 'ShortLink';
  use WWW::Shorten 'SmLnk';
  use WWW::Shorten 'URLjr';
  use WWW::Shorten 'V3';

  # Individual modules have have their
  # own syntactic variations.

  # See the documentation for the particular
  # module you intend to use for details, trips
  # and traps.

  $short_url = makeashorterlink($long_url);

  $long_url  = makealongerlink($short_url);

  # If you don't like the function names:
  use WWW::Shorten 'SnipURL', ':short';
  $short_url = short_link( $long_url );
  $long_url = long_link( $short_url );

=head1 ABSTRACT

A Perl interface to URL shortening sites. These sites maintain
databases of long URLs, each of which has a unique identifier.

=head1 DESCRIPTION

The function C<makeashorterlink> will call the relevant web site
passing it your long URL and will return the shorter version.

The function C<makealongerlink> does the reverse. C<makealongerlink>
will accept as an argument either the full shortened URL or just the
identifier.

If anything goes wrong, then either function will return C<undef>.

=cut

package WWW::Shorten;

use 5.006;
use strict;
use warnings;

use base qw(WWW::Shorten::generic);
our @EXPORT = qw(makeashorterlink makealongerlink);
our $VERSION = 1.98;

use Carp;

my $style;

=head1 Subroutines

=head2 import

Called when the module is C<use>d. Loads the correct sub-module

=cut

sub import
{
    my $class = shift;
    $style = shift;
    $style = 'Metamark' unless defined $style;
    my $package = "${class}::${style}";
    eval {
	my $file = $package;
	$file =~ s/::/\//g;
	require "$file.pm";
    };
    croak $@ if $@;
    $package->import( @_ );
}

1;

__END__

=head2 EXPORT

makeashorterlink, makealongerlink

Or, if you specify C<:short> on the import line, you instead
get C<short_link> and C<long_link>. If you explicitly want the
default set, use C<:default>.

Actually these functions are exported from the relevant subclass.

=head1 COMMAND LINE PROGRAM

A very simple program called F<shorten> is supplied in the
distribution's F<bin> folder. This program takes a URL and
gives you a shortened version of it.

=head1 EXAMPLES and PROGRAMS

Adam Kessel wrote F<shorlfilter>, a program that uses C<WWW::Shorten>
to filter a file, converting long URLs to short ones.

    http://bostoncoop.net/adam/shorlfilter

There is also a F<bin> directory in this distribution which contains a
sample program.

=head1 RELATED MODULES

=head2 Similar Aim

L<WWW::MakeAShorterLink> is what this module was originally based upon.
That module is obsoleted by L<WWW::Shorten::MakeAShorterLink>, one of
the many subclasses of L<WWW::Shorten>.

=head2 Same Area, Different Purpose

L<CGI::Shorten> provides building blocks for you to create your own URL
shortening service. It provides routines to shoretn a URL, lengthen one,
and it keeps a store. Nice and easy.

=head1 THANKS

Dave Cross for L<WWW::MakeAShorterLink>

Alex Page for the original LWP hacking on which Dave based his code.

Simon Batistoni for giving the C<makealongerlink> idea to Dave.

Eric Hammond for writing the NotLong variant.

Shashank Tripathi <shank@shank.com> for providing both SnipURL.com and
advice on the module.

Kevin Gilbertson (Gilby) supplied information on the TinyURL API.

Matt Felsen (mattf) wanted shorter function names.

Ask Bjoern Hansen for providing both Metamark.net and advice on the
module.

Martin Thurn for helping me notice a bug and for a suggestion regarding
F<MASL.pm>.

Jon and William (jon and wjr at smlnk.com respectively) for providing
SmLnk.com.

P J Goodwin for providing the code for L<WWW::Shorten::OneShortLink>.

And especial thanks to all providers of these services.

=head1 BUGS, REQUESTS, COMMENTS

Support for this module is supplied using the CPAN RT system via the web
or email:

    http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Shorten
    ( shorter URL: http://xrl.us/rfb )

    bug-www-shorten@rt.cpan.org

This makes it much easier for me to track things and thus means
your problem is less likely to be neglected.

=head1 LICENCE AND COPYRIGHT

WWW::Shorten::NotLong copyright (c) Eric Hammond <ehammond@thinksome.com>.

Other modules copyright (c) Magnum Solutions Ltd., 2007. All rights
reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.000 or,
at your option, any later version of Perl 5 you may have available.

The full text of the licences can be found in the F<Artistic> and
F<COPYING> files included with this module, or in L<perlartistic> and
L<perlgpl> as supplied with Perl 5.8.1 and later.

=head1 AUTHOR

Iain Truskett <spoon@cpan.org>

Based on WWW::MakeAShorterLink by Dave Cross <dave@mag-sol.com>

In 2004 Dave Cross took over the maintenance of this distribution
following the death of Iain Truskett.

=head1 SEE ALSO

L<perl>, L<CGI::Shorten>.

=cut
